### ISSUE TITLE
QSFetchZoneChangesOperation's ignoreDeviceIdentifier

### ISSUE THREAD

**Issue by BlixLT** [mentrena/SyncKit#77]
Hi, just wondering, why QSFetchZoneChangesOperation in `QSCloudKitSynchronizer.fetchZoneChanges:` method does not ignore self.deviceIdentifier, but it does in `QSCloudKitSynchronizer.updateServerTokenForRecordZones:` method. Is there a reason, why it shouldn't ignore it when fetching changes also? 
Because I noticed, that sometimes QSCloudKitSynchronizer download old changes with that device's identifier and apply it locally before sending new changes to cloudkit. And new data is being overridden with that old one. Is there a good reason, why changes from self.deviceIdentifier shouldn't be ignored when fetching changes? I cannot think of any, but I might be missing something obvious.

**Comment by mentrena** [comment]
Off the top of my head, I think it is needed to allow apps to use UICloudSharingController to share records. When you use UICloudSharingController, SyncKit creates and saves the CKShare for you, but UICloudSharingController can make changes to the CKShare by itself and those changes need to be downloaded (CKShare will still have that same device identifier)
Perhaps it would be possible to optimise the logic so records with the same device identifier are skipped _unless_ they're a share, I would need to investigate this.

> Because I noticed, that sometimes QSCloudKitSynchronizer download old changes with that device's identifier and apply it locally before sending new changes to cloudkit. And new data is being overridden with that old one

This puzzles me though, I would expect that your app wouldn't download changes that come from the same device, since the database/zone token would have been moved forward after uploading those changes. Do you have any idea if that happens often for you, or whether something in particular is causing this?

**Comment by BlixLT** [comment]
Ok, thanks for the answer. I tried sharing on mac OS only. But maybe something else was causing this, not that identifier. Not 100% sure. I will dig more into it when I have more time and let you know if I find something wrong.

**Comment by mentrena** [comment]
Thanks :)

**Comment by BlixLT** [comment]
Ok, here goes an update / question regarding this issue. I tried your sample iOS project and it seems that there is a bug when updating zone change tokens (unless I misunderstood logic somewhere). 

I created one company with three employees. Deleted one of them (in CloudKit dashboard). Then shared that company. Now every time I make a change (lets say rename some employee, add new one (even in not shared company)), then during updateServerTokenForRecordZones: I always get the same result:
```                
DLog(@"result.downloadedRecords: %@, result.deletedRecordIDs: %@", result.downloadedRecords, result.deletedRecordIDs);
                if (result.downloadedRecords.count || result.deletedRecordIDs.count) {
```
```
result.downloadedRecords: (
    "<CKShare: 0x7f9bb6e21550; values={\n    QSCloudKitDeviceUUIDKey = \"....\";\n}, participants=(\n    \"<CKShareParticipant: 0x600000a012d0; ....}>>\"\n), ...., recordType=cloudkit.share, etag=8>"
), result.deletedRecordIDs: (
    "<CKRecordID: 0x6000021a4940; recordName=QSEmployee.B652D859-DC43-48F1-B1C0-BBD866BF26ED, zoneID=QSCloudKitCustomZoneName:__defaultOwner__>"
)
```
Some time ago deleted employee is always being returned and any created share(s) are there in the downloadedResults. The cause of this seems to be that self.activeZoneTokens is empty in the updateServerTokenForRecordZones: method, because resetActiveTokens  was called somewhere earlier - as far as I can tell in the fetchDatabaseChangesWithCompletion: method. 
So, in the updateServerTokenForRecordZones: an empty self.activeZoneTokens dictionary is being provided to QSFetchZoneChangesOperation and therefore all changes in that zone being checked (since its creation and not from some previous server change token) and it always found deletions and CKShare changes that were made sometime earlier. Because of that change token is not being updated correctly and old change token is being given when fetching changes and recently uploaded change from the current device is being returned from cloudkit. I suppose this shouldn't happen.

Let me know if you need any additional info or something else for me to try.

**Comment by BlixLT** [comment]
Have you had a chance to look into this? Any news? Is it possible to fix (I think it does not work as expected). Or am I misunderstanding something?

**Comment by mentrena** [comment]
Hey! Not just yet, but I'll get to this soon.

**Comment by mentrena** [comment]
Just released 0.7.3 with a fix for this. Your message was really helpful, the problem was just as you described, thanks!
